nengo_spa/compiler/tests/test_parsing.py

```diff
@@ -0,0 +1,143 @@
+from io import BytesIO
+
+import pytest
+
+from nengo_spa.internal.generators import Peekable
+from nengo_spa.compiler import parsing
+from nengo_spa.compiler.parsing import MatchToken
+import nengo_spa.compiler.tokens as tk
+
+
+encoding_terminal = parsing.Terminal(type=tk.ENCODING, string='utf-8')
+
+
+def peekable_tokens(code):
+    tokens = Peekable(tk.tokenize(BytesIO(code.encode()).readline))
+    if encoding_terminal.accept(tokens):
+        encoding_terminal.read(tokens)
+    return tokens
+
+
+def test_terminal():
+    terminal = parsing.Terminal(type=tk.NAME, string='foobar')
+    tokens = peekable_tokens('foobar')
+    assert terminal.accept(tokens)
+    assert terminal.read(tokens) == [
+        MatchToken(type=tk.NAME, string='foobar')]
+
+    assert not terminal.accept(tokens)
+    with pytest.raises(parsing.RuleMismatchError):
+        terminal.read(tokens)
+
+
+@pytest.mark.parametrize('rule', [
+    parsing.Either(
+        parsing.Terminal(type=tk.NAME, string='foo'),
+        parsing.Terminal(type=tk.NAME, string='bar')),
+    parsing.Terminal(type=tk.NAME, string='foo') |
+    parsing.Terminal(type=tk.NAME, string='bar')])
+def test_either(rule):
+    tokens = peekable_tokens('foo bar')
+    assert rule.accept(tokens)
+    assert rule.read(tokens) == [
+        MatchToken(type=tk.NAME, string='foo')]
+
+    assert rule.accept(tokens)
+    assert rule.read(tokens) == [
+        MatchToken(type=tk.NAME, string='bar')]
+
+    assert not rule.accept(tokens)
+    with pytest.raises(parsing.RuleMismatchError):
+        rule.read(tokens)
+
+
+def test_either_chaining_gathers_rules_in_single_object():
+    rules = [parsing.Terminal() for i in range(3)]
+    either = rules[0] | rules[1] | rules[2]
+    assert all(r1 is r2 for r1, r2 in zip(either.rules, rules))
+
+
+@pytest.mark.parametrize('rule', [
+    parsing.Chain(
+        parsing.Terminal(type=tk.NAME, string='foo'),
+        parsing.Terminal(type=tk.NAME, string='bar')),
+    parsing.Terminal(type=tk.NAME, string='foo') +
+    parsing.Terminal(type=tk.NAME, string='bar')])
+def test_chain(rule):
+    tokens = peekable_tokens('foo bar')
+    assert rule.accept(tokens)
+    assert rule.read(tokens) == [
+        MatchToken(type=tk.NAME, string='foo'),
+        MatchToken(type=tk.NAME, string='bar')]
+
+    assert not rule.accept(tokens)
+    with pytest.raises(parsing.RuleMismatchError):
+        rule.read(tokens)
+
+
+def test_chaining_gathers_rules_in_single_object():
+    rules = [parsing.Terminal() for i in range(3)]
+    either = rules[0] + rules[1] + rules[2]
```

Comment:
Should this be called `chain` instead of `either` assuming this test is distinct from `test_either_chaining_gathers_rules_in_single_object`.